compatibility with fastn-stack/fastn/pull/1702

siddhantk232 opened the pull request.

siddhantk232 committed:
compatibility with fastn-stack/fastn/pull/1702

siddhantk232 committed:
Merge branch 'main' of github.com:fastn-stack/fastn_py into fix/name_changes

siddhantk232 committed:
format with black

amitu commented on fastn/django/__init__.py:
```diff
@@ -110,13 +109,12 @@ def _add_user(self, request: RequestType):
-            username=fastn_user.get("login"),
+            username=fastn_user.get("username"),
             defaults={
                 "first_name": first_name,
                 "last_name": last_name,
-                "email": fastn_user.get("email") or "",
-                # email has Not Null constraint
-            }
+                "email": fastn_user.get("email"),
+            },
         )
```
Let's convert this to a hook, 

```py
import django.utils
from django.conf import settings

django.utils.module_loading(
    settings.get_attr(
        "FASTN_AUTH_CALLBACK", 
        "fastn.django.default_callback"
    )
)(request=request, fastn_user=fastn_user)
```

Where `fastn.django.default_callback` will be some code like this:

```py
def default_callback(request, fastn_user):
       user, _ = User.objects.get_or_create(
            username=fastn_user.get("login"),
            username=fastn_user.get("username"),
            defaults={
                "first_name": first_name,
                "last_name": last_name,
                "email": fastn_user.get("email") or "",
                # email has Not Null constraint
            }
                "email": fastn_user.get("email"),
            },
        )
```

siddhantk232 replied:
should I also add call to `django.contrib.auth.login` in `default_callback`?

amitu replied:
Not sure. I think we will always need `auth.login()`.

amitu replied:
What does it do?

siddhantk232 replied:
it creates a django session entry for the authenticated user.

>  Persist a user id and a backend in the request. This way a user doesn't
    have to reauthenticate on every request. Note that data set during
    the anonymous session is retained when the user logs in.

siddhantk232 replied:
if we choose to move `auth.login` inside the callback. Our middleware will just be responsible for decoding the fastn set cookie into a python dict and then the callback can decide to do whatever it likes with it which I think is a better way.

siddhantk232 replied:
we can also get rid of the utility function `utils.get_first_name_and_last_name`

amitu replied:
Let's do this.

siddhantk232 replied:
done :)

siddhantk232 committed:
default_callback setting

a configurable callback function that will be called with the decoded
user from fastn set cookie.

A default implementation is also provided that creates the django USer
and `login()` the user

siddhantk232 committed:
doc: FASTN_AUTH_CALLBACK setting in README

amitu commented on fastn/django/__init__.py:
```diff
@@ -17,6 +18,10 @@
 
+FASTN_AUTH_CALLBACK = getattr(
+    settings, "FASTN_AUTH_CALLBACK", "fastn.django.default_callback"
+)
```
You can do the `module_loading.import_string` resolution part here as well to only do it once, instead on per request.

siddhantk232 replied:
[done](https://github.com/fastn-stack/fastn_py/pull/1/commits/1e414934204bf04152314daf3dcc7e7c811a3ce6)

siddhantk232 committed:
call module_loading.import_string on module load

amitu commented on fastn/django/__init__.py:
```diff
@@ -31,32 +35,32 @@ def wrapper(request: RequestType):
-    return django.http.JsonResponse({
-        "redirect": location,
-    })
+    return django.http.JsonResponse(
+        {
+            "redirect": location,
+        }
+    )
```
There is a redirect helper we can use.

amitu commented on fastn/django/__init__.py:
```diff
@@ -31,32 +35,32 @@ def wrapper(request: RequestType):
-    return django.http.JsonResponse({
-        "reload": True,
-    })
+    return django.http.JsonResponse(
+        {
+            "reload": True,
+        }
+    )
```
This should also use `reload()` helper.
